test/Microsoft.AspNetCore.Rewrite.Tests/UrlRewrite/FileParserTests.cs

```diff
@@ -0,0 +1,217 @@
+                IMurl: "^article/([0-9]+)/([_0-9a-z-]+)",
+                name: "Rewrite to article.aspx",
+                UAactionType: ActionType.Rewrite,
+                UApattern: "article.aspx?id={R:1}&amp;title={R:2}"));
+            expected.Add(CreateTestRule(condList,
+                IMurl: "^article/([0-9]+)/([_0-9a-z-]+)",
+                name: "Rewrite to article.aspx",
+                UAactionType: ActionType.Redirect,
+                UApattern: "article.aspx?id={R:1}&amp;title={R:2}"));
+
+            // act
+            var res = UrlRewriteFileParser.Parse(new StringReader(xml));
+
+            // assert
+            AssertUrlRewriteRuleEquality(expected, res);
+        }
+
+
+        // Creates a rule with appropriate default values of the url rewrite rule.
+        private UrlRewriteRule CreateTestRule(List<Condition> conditions,
+            LogicalGrouping condGrouping = LogicalGrouping.MatchAll,
+            bool condTracking = false,
+            string name = "",
+            bool enabled = true,
+            PatternSyntax patternSyntax = PatternSyntax.ECMAScript,
+            bool stopProcessing = false,
+            string IMurl = "",
+            bool IMignoreCase = true,
+            bool IMnegate = false,
+            ActionType UAactionType = ActionType.None,
+            string UApattern = "",
+            bool UAappendQueryString = false,
+            bool UAlogRewrittenUrl = false,
+            RedirectType UAredirectType = RedirectType.Permanent
+            )
+        {
+            return new UrlRewriteRule
+            {
+                Action = new UrlAction
+                {
+                    Url = InputParser.ParseInputString(UApattern),
+                    Type = UAactionType,
+                    AppendQueryString = UAappendQueryString,
+                    LogRewrittenUrl = UAlogRewrittenUrl,
+                    RedirectType = UAredirectType
+                },
+                Name = name,
+                Enabled = enabled,
+                StopProcessing = stopProcessing,
+                PatternSyntax = patternSyntax,
+                Match = new InitialMatch
+                {
+                    Url = new Regex(IMurl),
+                    IgnoreCase = IMignoreCase,
+                    Negate = IMnegate
+                },
+                Conditions = new Conditions
+                {
+                    ConditionList = conditions,
+                    MatchType = condGrouping,
+                    TrackingAllCaptures = condTracking
+                }
+            };
+        }
+
+        private void AssertUrlRewriteRuleEquality(List<UrlRewriteRule> expected, List<UrlRewriteRule> actual)
+        {
+
+            Assert.Equal(expected.Count, actual.Count);
+            for (var i = 0; i < expected.Count; i++)
+            {
+                var r1 = expected[i];
+                var r2 = actual[i];
+
+                Assert.Equal(r1.Name, r2.Name);
```

Comment:
Will we ever need to check for rule equality? Think we'd rather preserve reference equality on equals for rules.